### DISCUSSION TITLE
Approaches to avoid unnecessary DB reads

### DISCUSSION THREAD

**Discussion by Mesuva** [livewire/livewire#5493] [1. Help]
I'm using Livewire components to show effectively some TODO lists.
I have a pretty simple Livewire component for a single TODO item, which just has a title, due date and a status.

Each TODO item has a drop down of statuses for quickly changing them. The statuses themselves are pulled from a simple database table, since they might change a bit over time, wording, options, etc.

If I fetch the list of statuses in the render function of the TODO component, passing it to the view, that works fine.
But since this gets called every time there's a TODO item on a page, it's fetching from the DB the same data over and over within the same request. If a page has 40 TODO components on it, it's doing 80 DB requests (once for the TODO list data, and once for the list of statuses).

There's some good reasons I have individual todo items as their own components, as I sometimes show them individually, sometimes nest them within a large component that groups them. I did consider just making one larger TODO list with multiple items in it, but it then starts to get away from the re-usability of smaller components.. 
(and perhaps it's more the pattern I'm asking about here anyway)

Laravel's caching isn't going to help me, as I'm running serverless (Bref), and that means just swapping out the original request with a repeated request to the cache database table. This data just isn't going to change across a single page load, or even from week to week. I'm really just talking about within the same page request too.

So what are the recommended approaches here?
- Create a View Composer, and store the list of statuses on the class property, so it can be passed in from an already fetched value when the livewire blade template is fired up?  (Pretty sure a View Composer is loaded as a singleton)
- Create some sort of static function and property somewhere (perhaps on the model), and fetch from that anytime it's needed?
- Some sort of trick with passing down shared values to child components that I've perhaps not seen?
- Some other trick with caching or something else?

I absolutely love Livewire, but I often feel I overlook times where I'm hitting the database too often, even just on the same initial page load.

**Comment by joshhanley** [comment]
@Mesuva personally I would have one full page Livewire component that controls everything and use blade components to split up the UI into more manageable pieces. If possible, I would try to refactor it to this.

But if not, then for the initial render of the page, I would still have one wrapping Livewire component, that could load the statuses once (maybe using a computed property). Then you could pass the list of statuses from the wrapping component into your individual todo Livewire components, through the mount method.

Now one trick you could do with that, is assign the statuses to a protected property, instead of a public one. Then you can make use of a computed property to access the statuses from the view. Inside the computed property, it can check for any data in the protected statuses property and if so, then return them. But if the protected statuses property is null, then it could do a db query to retrieve them.

This would ensure that the statuses can be passed through mount and used on the initial render, and on any subsequent renders of individual todo components, they can then load the statuses from the db as needed.

Code would look something like this:

```php
TodoList extends Component
{
    public function getStatusesProperty()
    {
        return Status::all(); // Change to however you are loading statuses
    }
}
```

```blade
{{-- todo-list.blade.php --}}
<div>
    <livewire:todo-item :statusesList="$this->statuses" /> {{-- Access statuses computed property using `$this->` --}}
</div>
```

```php
TodoItem extends Component
{
    protected $_statuses; // This can't be called `$statuses` as the computed property will be called that.
    
    public function mount($statuses)
    {
        $this->_statuses = $statuses;
    }
    
    public function getStatusesProperty()
    {
        if ($this->_statuses) {
            return $this->_statuses;
        }
        
        return Status::all(); // Change to however you are loading statuses
    }
}
```

```blade
{{-- todo-item.blade.php --}}
<div>
    @foreach ($this->statuses as $status) {{-- Access statuses computed property using `$this->` --}}
        {{-- output your statuses as needed --}}
    @endforeach
</div>
```

Hope this helps!
